File: mkube-console/internal/cluster/aggregator.go
package cluster

import (
	"context"
	"fmt"
	"io"
	"log"
	"sync"
	"time"

	corev1 "k8s.io/api/core/v1"
	"golang.org/x/sync/errgroup"
)

// Aggregator manages multiple NodeClients and provides a unified cluster view.
type Aggregator struct {
	mu      sync.RWMutex
	clients map[string]*NodeClient // name -> client
}

// NewAggregator creates an aggregator with the given node clients.
func NewAggregator(clients []*NodeClient) *Aggregator {
	m := make(map[string]*NodeClient, len(clients))
	for _, c := range clients {
		m[c.Name] = c
	}
	return &Aggregator{clients: m}
}

// ClusterSummary contains aggregated cluster stats.
type ClusterSummary struct {
	NodeCount    int
	HealthyNodes int
	PodCount     int
	RunningPods  int
	Nodes        []NodeSummary
}

// NodeSummary contains stats for one node.
type NodeSummary struct {
	Name      string
	Healthy   bool
	PodCount  int
	LastPing  time.Time
}

// ListAllPods returns pods from all nodes, annotated with mkube.io/node.
func (a *Aggregator) ListAllPods(ctx context.Context) ([]corev1.Pod, error) {
	a.mu.RLock()
	clients := make([]*NodeClient, 0, len(a.clients))
	for _, c := range a.clients {
		clients = append(clients, c)
	}
	a.mu.RUnlock()

	type result struct {
		pods []corev1.Pod
		node string
	}

	var mu sync.Mutex
	var allPods []corev1.Pod

	g, ctx := errgroup.WithContext(ctx)
	for _, client := range clients {
		c := client
		g.Go(func() error {
			list, err := c.ListPods(ctx)
			if err != nil {
				log.Printf("error listing pods from %s: %v", c.Name, err)
				return nil // don't fail the whole operation
			}
			mu.Lock()
			for i := range list.Items {
				pod := list.Items[i]
				if pod.Annotations == nil {
					pod.Annotations = make(map[string]string)
				}
				pod.Annotations["mkube.io/node"] = c.Name
				allPods = append(allPods, pod)
			}
			mu.Unlock()
			return nil
		})
	}
	_ = g.Wait()
	return allPods, nil
}

// ListAllNodes returns node objects from all clients.
func (a *Aggregator) ListAllNodes(ctx context.Context) ([]corev1.Node, error) {
	a.mu.RLock()
	clients := make([]*NodeClient, 0, len(a.clients))
	for _, c := range a.clients {
		clients = append(clients, c)
	}
	a.mu.RUnlock()

	var mu sync.Mutex
	var nodes []corev1.Node

	g, ctx := errgroup.WithContext(ctx)
	for _, client := range clients {
		c := client
		g.Go(func() error {
			node, err := c.GetNode(ctx)
			if err != nil {
				log.Printf("error getting node from %s: %v", c.Name, err)
				return nil
			}
			mu.Lock()
			nodes = append(nodes, *node)
			mu.Unlock()
			return nil
		})
	}
	_ = g.Wait()
	return nodes, nil
}

// GetPod searches all nodes for the given pod.
func (a *Aggregator) GetPod(ctx context.Context, ns, name string) (*corev1.Pod, string, error) {
	a.mu.RLock()
	clients := make([]*NodeClient, 0, len(a.clients))
	for _, c := range a.clients {
		clients = append(clients, c)
	}
	a.mu.RUnlock()

	for _, c := range clients {
		pod, err := c.GetPod(ctx, ns, name)
		if err == nil {
			if pod.Annotations == nil {
				pod.Annotations = make(map[string]string)
			}
			pod.Annotations["mkube.io/node"] = c.Name
			return pod, c.Name, nil
		}
	}
	return nil, "", fmt.Errorf("pod %s/%s not found on any node", ns, name)
}

// CreatePod routes the pod to the appropriate node.
// If spec.nodeName is set, use that node. Otherwise use the node with fewest pods.
func (a *Aggregator) CreatePod(ctx context.Context, pod *corev1.Pod) (*corev1.Pod, error) {
	a.mu.RLock()
	defer a.mu.RUnlock()

	var target *NodeClient

	// Route by nodeName if specified
	if pod.Spec.NodeName != "" {
		if c, ok := a.clients[pod.Spec.NodeName]; ok {
			target = c
		} else {
			return nil, fmt.Errorf("node %q not found", pod.Spec.NodeName)
		}
	}

	// Least-pods scheduling
	if target == nil {
		minPods := int(^uint(0) >> 1)
		for _, c := range a.clients {
			if !c.IsHealthy() {
				continue
			}
			list, err := c.ListPods(ctx)
			if err != nil {
				continue
			}
			if len(list.Items) < minPods {
				minPods = len(list.Items)
				target = c
			}
		}
	}

	if target == nil {
		return nil, fmt.Errorf("no healthy nodes available")
	}

	return target.CreatePod(ctx, pod)
}

// DeletePod finds the owning node and deletes the pod.
func (a *Aggregator) DeletePod(ctx context.Context, ns, name string) error {
	_, nodeName, err := a.GetPod(ctx, ns, name)
	if err != nil {
		return err
	}

	a.mu.RLock()
	c, ok := a.clients[nodeName]
	a.mu.RUnlock()
	if !ok {
		return fmt.Errorf("node %q not found", nodeName)
	}

	return c.DeletePod(ctx, ns, name)
}

// GetPodLog proxies to the owning node.
func (a *Aggregator) GetPodLog(ctx context.Context, ns, name string) (io.ReadCloser, error) {
	_, nodeName, err := a.GetPod(ctx, ns, name)
	if err != nil {
		return nil, err
	}

	a.mu.RLock()
	c, ok := a.clients[nodeName]
	a.mu.RUnlock()
	if !ok {
		return nil, fmt.Errorf("node %q not found", nodeName)
	}

	return c.GetPodLog(ctx, ns, name)
}

// GetClusterSummary builds an aggregate summary of the cluster.
func (a *Aggregator) GetClusterSummary(ctx context.Context) *ClusterSummary {
	a.mu.RLock()
	clients := make([]*NodeClient, 0, len(a.clients))
	for _, c := range a.clients {
		clients = append(clients, c)
	}
	a.mu.RUnlock()

	summary := &ClusterSummary{
		NodeCount: len(clients),
	}

	for _, c := range clients {
		ns := NodeSummary{
			Name:     c.Name,
			Healthy:  c.IsHealthy(),
			LastPing: c.LastPing(),
		}
		if c.IsHealthy() {
			summary.HealthyNodes++
		}
		if list, err := c.ListPods(ctx); err == nil {
			ns.PodCount = len(list.Items)
			summary.PodCount += len(list.Items)
			for _, pod := range list.Items {
				if pod.Status.Phase == corev1.PodRunning {
					summary.RunningPods++
				}
			}
		}
		summary.Nodes = append(summary.Nodes, ns)
	}

	return summary
}

// RunHealthChecker periodically pings all nodes.
func (a *Aggregator) RunHealthChecker(ctx context.Context) {
	ticker := time.NewTicker(15 * time.Second)
	defer ticker.Stop()

	// Initial check
	a.pingAll(ctx)

	for {
		select {
		case <-ctx.Done():
			return
		case <-ticker.C:
			a.pingAll(ctx)
		}
	}
}

func (a *Aggregator) pingAll(ctx context.Context) {
	a.mu.RLock()
	clients := make([]*NodeClient, 0, len(a.clients))
	for _, c := range a.clients {
		clients = append(clients, c)
	}
	a.mu.RUnlock()

	for _, c := range clients {
		if err := c.Ping(ctx); err != nil {
			log.Printf("health check failed for %s: %v", c.Name, err)
		}
	}
}

// GetNode returns a node by name.
func (a *Aggregator) GetNode(ctx context.Context, name string) (*corev1.Node, error) {
	a.mu.RLock()
	c, ok := a.clients[name]
	a.mu.RUnlock()
	if !ok {
		return nil, fmt.Errorf("node %q not found", name)
	}
	return c.GetNode(ctx)
}

File: src/clients/aggregator.rs
use std::collections::HashMap;
use std::sync::Arc;
use tokio::sync::RwLock;
use tokio::time::{self, Duration};
use tracing::{info, warn};

use crate::models::k8s::{Node, Pod};
use crate::models::views::{ClusterSummary, NodeSummary};

use super::NodeClient;

pub struct Aggregator {
    clients: RwLock<HashMap<String, Arc<NodeClient>>>,
}

impl Aggregator {
    pub fn new(clients: Vec<NodeClient>) -> Self {
        let mut m = HashMap::new();
        for c in clients {
            m.insert(c.name.clone(), Arc::new(c));
        }
        Self {
            clients: RwLock::new(m),
        }
    }

    pub async fn list_all_pods(&self) -> Result<Vec<Pod>, Box<dyn std::error::Error + Send + Sync>> {
        let clients = self.snapshot().await;

        let mut all_pods = Vec::new();
        let mut handles = Vec::new();

        for client in clients {
            let c = client.clone();
            handles.push(tokio::spawn(async move {
                match c.list_pods().await {
                    Ok(list) => Some((c.name.clone(), list)),
                    Err(e) => {
                        warn!("error listing pods from {}: {}", c.name, e);
                        None
                    }
                }
            }));
        }

        for handle in handles {
            if let Ok(Some((node_name, list))) = handle.await {
                for mut pod in list.items {
                    let annotations = pod.metadata.annotations.get_or_insert_with(HashMap::new);
                    annotations.insert("mkube.io/node".to_string(), node_name.clone());
                    all_pods.push(pod);
                }
            }
        }

        Ok(all_pods)
    }

    pub async fn list_all_nodes(
        &self,
    ) -> Result<Vec<Node>, Box<dyn std::error::Error + Send + Sync>> {
        let clients = self.snapshot().await;

        let mut nodes = Vec::new();
        let mut handles = Vec::new();

        for client in clients {
            let c = client.clone();
            handles.push(tokio::spawn(async move {
                match c.get_node().await {
                    Ok(node) => Some(node),
                    Err(e) => {
                        warn!("error getting node from {}: {}", c.name, e);
                        None
                    }
                }
            }));
        }

        for handle in handles {
            if let Ok(Some(node)) = handle.await {
                nodes.push(node);
            }
        }

        Ok(nodes)
    }

    pub async fn get_pod(
        &self,
        ns: &str,
        name: &str,
    ) -> Result<(Pod, String), Box<dyn std::error::Error + Send + Sync>> {
        let clients = self.snapshot().await;

        for client in &clients {
            if let Ok(mut pod) = client.get_pod(ns, name).await {
                let annotations = pod.metadata.annotations.get_or_insert_with(HashMap::new);
                annotations.insert("mkube.io/node".to_string(), client.name.clone());
                return Ok((pod, client.name.clone()));
            }
        }
        Err(format!("pod {}/{} not found on any node", ns, name).into())
    }

    pub async fn create_pod(
        &self,
        pod: &Pod,
    ) -> Result<Pod, Box<dyn std::error::Error + Send + Sync>> {
        let clients_map = self.clients.read().await;

        // Route by nodeName if specified
        if !pod.spec.node_name.is_empty() {
            if let Some(c) = clients_map.get(&pod.spec.node_name) {
                return c.create_pod(pod).await;
            }
            return Err(format!("node {:?} not found", pod.spec.node_name).into());
        }

        // Least-pods scheduling
        let mut target: Option<Arc<NodeClient>> = None;
        let mut min_pods = usize::MAX;

        for c in clients_map.values() {
            if !c.is_healthy() {
                continue;
            }
            if let Ok(list) = c.list_pods().await {
                if list.items.len() < min_pods {
                    min_pods = list.items.len();
                    target = Some(c.clone());
                }
            }
        }

        match target {
            Some(c) => c.create_pod(pod).await,
            None => Err("no healthy nodes available".into()),
        }
    }

    pub async fn delete_pod(
        &self,
        ns: &str,
        name: &str,
    ) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
        let (_, node_name) = self.get_pod(ns, name).await?;

        let clients_map = self.clients.read().await;
        let c = clients_map
            .get(&node_name)
            .ok_or_else(|| format!("node {:?} not found", node_name))?;
        c.delete_pod(ns, name).await
    }

    pub async fn get_pod_log(
        &self,
        ns: &str,
        name: &str,
    ) -> Result<String, Box<dyn std::error::Error + Send + Sync>> {
        let (_, node_name) = self.get_pod(ns, name).await?;

        let clients_map = self.clients.read().await;
        let c = clients_map
            .get(&node_name)
            .ok_or_else(|| format!("node {:?} not found", node_name))?;
        c.get_pod_log(ns, name).await
    }

    pub async fn get_node(
        &self,
        name: &str,
    ) -> Result<Node, Box<dyn std::error::Error + Send + Sync>> {
        let clients_map = self.clients.read().await;
        let c = clients_map
            .get(name)
            .ok_or_else(|| format!("node {:?} not found", name))?;
        c.get_node().await
    }

    pub async fn get_cluster_summary(&self) -> ClusterSummary {
        let clients = self.snapshot().await;

        let mut summary = ClusterSummary {
            node_count: clients.len(),
            ..Default::default()
        };

        for c in &clients {
            let mut ns = NodeSummary {
                name: c.name.clone(),
                healthy: c.is_healthy(),
                pod_count: 0,
                last_ping: c.last_ping(),
            };

            if c.is_healthy() {
                summary.healthy_nodes += 1;
            }

            if let Ok(list) = c.list_pods().await {
                ns.pod_count = list.items.len();
                summary.pod_count += list.items.len();
                for pod in &list.items {
                    if pod.status.phase == "Running" {
                        summary.running_pods += 1;
                    }
                }
            }

            summary.nodes.push(ns);
        }

        summary
    }

    pub async fn run_health_checker(self: Arc<Self>, mut shutdown: tokio::sync::watch::Receiver<()>) {
        // Initial check
        self.ping_all().await;

        let mut interval = time::interval(Duration::from_secs(15));
        interval.tick().await; // skip first immediate tick

        loop {
            tokio::select! {
                _ = interval.tick() => {
                    self.ping_all().await;
                }
                _ = shutdown.changed() => {
                    info!("health checker shutting down");
                    return;
                }
            }
        }
    }

    async fn ping_all(&self) {
        let clients = self.snapshot().await;
        for c in &clients {
            if let Err(e) = c.ping().await {
                warn!("health check failed for {}: {}", c.name, e);
            }
        }
    }

    async fn snapshot(&self) -> Vec<Arc<NodeClient>> {
        self.clients.read().await.values().cloned().collect()
    }
}
